src/network/networkd-wwan.c

```diff
@@ -0,0 +1,666 @@
+                        if (r < 0)
+                                ret = log_link_warning_errno(link, r, "Could not stop IPv6 Router Discovery: %m");
+                }
+        }
+
+        /* Finally, remove all marked configs. */
+        SET_FOREACH(address, link->addresses) {
+                if (address->source != NETWORK_CONFIG_SOURCE_MODEM_MANAGER)
+                        continue;
+
+                if (!address_is_marked(address))
+                        continue;
+
+                r = address_remove(address, link);
+                if (r < 0)
+                        ret = r;
+        }
+
+        SET_FOREACH(route, link->manager->routes) {
+                if (route->source != NETWORK_CONFIG_SOURCE_MODEM_MANAGER)
+                        continue;
+
+                if (!route_is_marked(route))
+                        continue;
+
+                r = route_remove(route, link->manager);
+                if (ret)
+                        ret = r;
+        }
+
+        if (ret < 0)
+                return ret;
+
+        if (link->bearer_messages == 0)
+                link->bearer_configured = true;
+
+        if (!link->bearer_configured)
+                link_set_state(link, LINK_STATE_CONFIGURING);
+
+        link_check_ready(link);
+
+        return ret;
+}
+
+int link_apply_bearer(Link *link) {
+        Bearer *b = NULL;
+        int r;
+
+        assert(link);
+
+        (void) link_get_bearer(link, &b);
+
+        r = link_apply_bearer_impl(link, b);
+        if (r < 0)
+                link_enter_failed(link);
+
+        return r;
+}
+
+int bearer_update_link(Bearer *b) {
+        Link *link;
+        int r;
+
+        assert(b);
+        assert(b->modem);
+        assert(b->modem->manager);
+
+        if (!b->name)
+                return 0;
+
+        if (link_get_by_name(b->modem->manager, b->name, &link) < 0)
+                return 0;
+
+        r = link_reconfigure_impl(link, 0);
+        if (r < 0)
+                link_enter_failed(link);
+        if (r != 0) /* r > 0 means interface is reconfigured. */
+                return r;
+
+        r = link_apply_bearer_impl(link, b);
+        if (r < 0)
+                link_enter_failed(link);
+
+        /*
+         * Need to bring up the interface after the modem has connected.
+         * This is because ModemManger does the following while connecting:
+         * <msg> [1755871777.322239] [modem2] state changed (registered -> connecting)
+         * <dbg> [1755871777.325012] [modem2/bearer5] launching connection with QMI port (cdc-wdm0) and data port (wwan0) (multiplex none)
+         * <dbg> [1755871777.327665] [cdc-wdm0/qmi] bringing down data interface 'wwan0'
+         * <dbg> [1755871777.330108] [modem2/wwan0/net] interface index: 9
+         * <dbg> [1755871777.335265] [cdc-wdm0/qmi] deleting all links in data interface 'wwan0'
+         */
+
+        r = link_request_to_bring_up_or_down(link, b->connected);
```

Comment:
Hmm, please do not call here. If the ordering is important, please adjust link_is_ready_to_bring_up_or_down().

Reply:
So, I may misunderstand the situation, but maybe better to not bring up the interface at the beginning (by modifying `link_is_ready_to_bring_up_or_down()`), then after MM establish the connection, we should bring up the interface if necessary.

Reply:
Hmm, I will reconsider this later...

Reply:
If ModemManager bring down the interface on connect, maybe better to check the bearer state in `link_admin_state_down()` or `link_carrier_lost()`?

Reply:
Well, it seems the proposed functions are called too early for that:

```
wwan0: Link DOWN
XXXXXXXXXXXXXXXXXX link_admin_state_down
wwan0: Lost carrier
XXXXXXXXXXXXXXXXXX link_carrier_lost
XXXXXXXXXXXXXXXXXX bearer_update_link enter
wwan0: Flags change: -BROADCAST +POINTOPOINT +NOARP
ModemManager: Quectel EM12-G is connected, interface wwan0
XXXXXXXXXXXXXXXXXX bearer_update_link enter
wwan0: link_check_ready(): Bearer has not been applied.
XXXXXXXXXXXXXXXXXX bearer_update_link: connected 1

```
This is what I have from the real run, so I would suggest we keep the existing link_request_to_bring_up_or_down call where it is.

Reply:
OK. Let's keep it at least for a while. I will reconsider later.